org/mozilla/jss/pkcs11/PK11Signature.c

```diff
@@ -48,39 +55,79 @@ Java_org_mozilla_jss_pkcs11_PK11Signature_initSigContext
     SGNContext *ctxt=NULL;
     jobject contextProxy=NULL;
     SECKEYPrivateKey *privk;
+    SECAlgorithmID *signAlg=NULL;
+    SECStatus rv;
+    int keyTypeOKForPSS = 0;
 
     /* Extract the private key from the PK11Signature */
     if( getPrivateKey(env, this, &privk) != PR_SUCCESS) {
         PR_ASSERT( (*env)->ExceptionOccurred(env) != NULL);
         goto finish;
     }
+    
+    // See if the key is a candidate for PSS signature 
+    if (privk->keyType == rsaKey || privk->keyType == rsaPssKey) {
+        keyTypeOKForPSS = 1;
+    }
+    
+    PRArenaPool *arena = NULL;
+    SECOidTag signingAlg = SEC_OID_UNKNOWN;
 
-    /* Start the signing operation */
-    ctxt = SGN_NewContext(getAlgorithm(env, this), privk);
-	if(ctxt == NULL) {
-		JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to create signing context");
-		goto finish;
-	}
+    signingAlg = getAlgorithm(env,this);
+
+    if(signingAlg == SEC_OID_PKCS1_RSA_PSS_SIGNATURE && keyTypeOKForPSS) {
+        arena = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+        if (!arena) {
+           JSS_throw(env, OUT_OF_MEMORY_ERROR);
+           goto finish;
+        }
+
+        signAlg = getRSAPSSParamsAndSigningAlg(arena,env, this,privk);
+
+        if (signAlg == NULL) {
+            goto finish;
+        }
+
+        /* Start the signing operation */
+        ctxt = SGN_NewContextWithAlgorithmID(signAlg, privk);
+
+    } else {
+        ctxt = SGN_NewContext(signingAlg, privk);
+    }
+
+    if(ctxt == NULL) {
+        if(arena != NULL) {
+            PORT_FreeArena(arena, PR_FALSE);
+        }
+
+        JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to create signing context");
+	goto finish;
+    }
     if( SGN_Begin(ctxt) != SECSuccess ) {
         JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to begin signing context");
         goto finish;
     }
 
     /* Create a contextProxy and stick it in the PK11Signature object */
     contextProxy = JSS_PK11_wrapSigContextProxy(env,
-												(void**)&ctxt,
-												SGN_CONTEXT);
+        (void**)&ctxt,
+        SGN_CONTEXT,arena);
+
     if(contextProxy == NULL) {
         PR_ASSERT( (*env)->ExceptionOccurred(env) != NULL);
         goto finish;
     }
+
     setSigContext(env, this, contextProxy);
 
 finish:
```

Comment:
FWIW, I think you leak `signAlg` here. If you accept my [other suggestion](https://github.com/dogtagpki/jss/pull/445/files#r398823299), you wouldn't.

Reply:
Ah I see. You're doing an arena-based allocation for the `signAlg` too. That definitely warrants a comment.

Reply:
I will look at it, I did do a valgrind run at one point and it looked ok, thx...

Reply:
Added comment somewhere :)